On some projects "CollectDeclaredReferencesTask" throws System.NullReferenceException

I have a C# solution and I added ReferenceTrimmer nuget package to all the projects. 3 out of 9 projects give following kind error when compiling. When removing ReferenceTrimmer from those projects they compile fine. I can't think of anything common with these three projects that other projects don't have. 

> ------ Build started: Project: StickerPrinting, Configuration: Debug x64 ------
C:\Users\bantza\.nuget\packages\referencetrimmer\3.1.22\build\ReferenceTrimmer.targets(24,5): error MSB4018: The "CollectDeclaredReferencesTask" task failed unexpectedly.
&nbsp;&nbsp;System.NullReferenceException: Object reference not set to an instance of an object.
&nbsp;&nbsp;&nbsp;&nbsp;at ReferenceTrimmer.Tasks.CollectDeclaredReferencesTask.GetPackageInfos()
&nbsp;&nbsp;&nbsp;&nbsp;at ReferenceTrimmer.Tasks.CollectDeclaredReferencesTask.Execute()
&nbsp;&nbsp;&nbsp;&nbsp;at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
&nbsp;&nbsp;&nbsp;&nbsp;at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()

I'm using VisualStudio.17.Release/17.7.3+34024.191 and the projects in the solution are .netframework4.8 and .netstandard2.0. All three that are failing are .netframework4.8, but there are also other .netframework4.8 projects that are not failing.

--- a/src/E2ETests/E2ETests.cs
+++ b/src/E2ETests/E2ETests.cs
@@ -240,6 +240,20 @@ public sealed class E2ETests
             expectedWarnings: Array.Empty<Warning>());
     }
 
+    [TestMethod]
+    public void LegacyStyleProject()
+    {
+        if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+        {
+            Assert.Inconclusive("This test only applies to Windows");
+            return;
+        }
+
+        RunMSBuild(
+            projectFile: "Library/Library.csproj",
+            expectedWarnings: Array.Empty<Warning>());
+    }
+
     private static (string ExePath, string Verb) GetMsBuildExeAndVerb()
     {
         // On Windows, try to find Visual Studio

--- a/src/E2ETests/TestData/LegacyStyleProject/Library/Class1.cs
+++ b/src/E2ETests/TestData/LegacyStyleProject/Library/Class1.cs
@@ -1,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ClassLibrary1
+{
+    public class Class1
+    {
+    }
+}

--- a/src/E2ETests/TestData/LegacyStyleProject/Library/Library.csproj
+++ b/src/E2ETests/TestData/LegacyStyleProject/Library/Library.csproj
@@ -1,0 +1,50 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
+  <PropertyGroup>
+    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
+    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <ProjectGuid>80472081-bc7f-43a9-af72-dc02c677c2c1</ProjectGuid>
+    <OutputType>Library</OutputType>
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <RootNamespace>Library</RootNamespace>
+    <AssemblyName>Library</AssemblyName>
+    <TargetFrameworkVersion>v4.7.2</TargetFrameworkVersion>
+    <FileAlignment>512</FileAlignment>
+    <Deterministic>true</Deterministic>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
+    <DebugSymbols>true</DebugSymbols>
+    <DebugType>full</DebugType>
+    <Optimize>false</Optimize>
+    <OutputPath>bin\Debug\</OutputPath>
+    <DefineConstants>DEBUG;TRACE</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
+    <DocumentationFile>bin\Debug\Library.xml</DocumentationFile>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+    <DebugType>pdbonly</DebugType>
+    <Optimize>true</Optimize>
+    <OutputPath>bin\Release\</OutputPath>
+    <DefineConstants>TRACE</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
+    <DocumentationFile>bin\Release\Library.xml</DocumentationFile>
+  </PropertyGroup>
+  <ItemGroup>
+    <Reference Include="System"/>
+    <Reference Include="System.Core"/>
+    <Reference Include="System.Xml.Linq"/>
+    <Reference Include="System.Data.DataSetExtensions"/>
+    <Reference Include="Microsoft.CSharp"/>
+    <Reference Include="System.Data"/>
+    <Reference Include="System.Net.Http"/>
+    <Reference Include="System.Xml"/>
+  </ItemGroup>
+  <ItemGroup>
+    <Compile Include="Class1.cs" />
+    <Compile Include="Properties\AssemblyInfo.cs" />
+  </ItemGroup>
+  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
+ </Project>

--- a/src/E2ETests/TestData/LegacyStyleProject/Library/Properties/AssemblyInfo.cs
+++ b/src/E2ETests/TestData/LegacyStyleProject/Library/Properties/AssemblyInfo.cs
@@ -1,0 +1,36 @@
+﻿using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("ClassLibrary1")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
+[assembly: AssemblyProduct("ClassLibrary1")]
+[assembly: AssemblyCopyright("Copyright ©  2023")]
+[assembly: AssemblyTrademark("")]
+[assembly: AssemblyCulture("")]
+
+// Setting ComVisible to false makes the types in this assembly not visible
+// to COM components.  If you need to access a type in this assembly from
+// COM, set the ComVisible attribute to true on that type.
+[assembly: ComVisible(false)]
+
+// The following GUID is for the ID of the typelib if this project is exposed to COM
+[assembly: Guid("80472081-bc7f-43a9-af72-dc02c677c2c1")]
+
+// Version information for an assembly consists of the following four values:
+//
+//      Major Version
+//      Minor Version
+//      Build Number
+//      Revision
+//
+// You can specify all the values or you can default the Build and Revision Numbers
+// by using the '*' as shown below:
+// [assembly: AssemblyVersion("1.0.*")]
+[assembly: AssemblyVersion("1.0.0.0")]
+[assembly: AssemblyFileVersion("1.0.0.0")]

--- a/src/Package/build/ReferenceTrimmer.targets
+++ b/src/Package/build/ReferenceTrimmer.targets
@@ -6,7 +6,7 @@
     <CoreCompileDependsOn Condition="'$(EnableReferenceTrimmer)' != 'false'">$(CoreCompileDependsOn);CollectDeclaredReferences</CoreCompileDependsOn>
   </PropertyGroup>
 
-  <Target Name="CollectDeclaredReferences" DependsOnTargets="ResolveProjectReferences">
+  <Target Name="CollectDeclaredReferences" DependsOnTargets="ResolveAssemblyReferences;PrepareProjectReferences">
     <PropertyGroup>
       <_ReferenceTrimmerDeclaredReferencesFile>$(IntermediateOutputPath)\_ReferenceTrimmer_DeclaredReferences.tsv</_ReferenceTrimmerDeclaredReferencesFile>
     </PropertyGroup>
@@ -28,7 +28,8 @@
       ProjectReferences="@(_ReferenceTrimmerProjectReferences)"
       PackageReferences="@(PackageReference)"
       ProjectAssetsFile="$(ProjectAssetsFile)"
-      TargetFramework="$(TargetFramework)"
+      TargetFrameworkMoniker="$(ReferringTargetFrameworkForProjectReferences)"
+      TargetPlatformMoniker="$(TargetPlatformMoniker)"
       RuntimeIdentifier="$(RuntimeIdentifier)"
       NuGetRestoreTargets="$(NuGetRestoreTargets)"
       TargetFrameworkDirectories="$(TargetFrameworkDirectory)"

--- a/src/Tasks/CollectDeclaredReferencesTask.cs
+++ b/src/Tasks/CollectDeclaredReferencesTask.cs
@@ -2,6 +2,7 @@ using System.Reflection;
 using System.Xml.Linq;
 using Microsoft.Build.Framework;
 using NuGet.Common;
+using NuGet.Frameworks;
 using NuGet.ProjectModel;
 using ReferenceTrimmer.Shared;
 using MSBuildTask = Microsoft.Build.Utilities.Task;
@@ -36,7 +37,9 @@ public sealed class CollectDeclaredReferencesTask : MSBuildTask
 
     public string? ProjectAssetsFile { get; set; }
 
-    public string? TargetFramework { get; set; }
+    public string? TargetFrameworkMoniker { get; set; }
+
+    public string? TargetPlatformMoniker { get; set; }
 
     public string? RuntimeIdentifier { get; set; }
 
@@ -185,7 +188,8 @@ public sealed class CollectDeclaredReferencesTask : MSBuildTask
         var lockFile = LockFileUtilities.GetLockFile(ProjectAssetsFile, NullLogger.Instance);
         var packageFolders = lockFile.PackageFolders.Select(item => item.Path).ToList();
 
-        var nugetTarget = lockFile.GetTarget(TargetFramework, RuntimeIdentifier);
+        var nugetFramework = NuGetFramework.ParseComponents(TargetFrameworkMoniker, TargetPlatformMoniker);
+        var nugetTarget = lockFile.GetTarget(nugetFramework, RuntimeIdentifier);
         var nugetLibraries = nugetTarget.Libraries
             .Where(nugetLibrary => nugetLibrary.Type.Equals("Package", StringComparison.OrdinalIgnoreCase))
             .ToList();

